### PR TITLE
[Backport ready] Fix bug when topic names are not in all lowercase

### PR DESCRIPTION
**Pull request by purbon**
As described in (https://github.com/confluentinc/kafka-connect-elasticsearch/issues/246#issuecomment-431794980), Elasticsearch can only create indexes in lowercase letters, this PR fix this by making sure that all indexes are created that way.

Please forgive me If I miss a necessary check before pushing the PR, is the first time I contribute a change. 

Fix partially #246 

Backport of #250 as requested in https://github.com/confluentinc/kafka-connect-elasticsearch/pull/250#pullrequestreview-167503044

### CODE REVIEW

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java [hunk 1]
```diff
@@ -72,4 +79,39 @@ public void testPutAndFlush() throws Exception {
...
+            key,
+            schema,
+            record,
+            0 );
+
+    try {
+      task.start(props, client);
+      task.open(new HashSet<>(Collections.singletonList(TOPIC_IN_CAPS_PARTITION)));
+      task.put(Collections.singleton(sinkRecord));
+      assertTrue("A topic name not in lowercase was created in Elasticsearch", true);
```
Comment (rhauch) [inline src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java]: This assertion doesn't really check anything, since you're always passing `true` into the method. Is this intentional?

Don't we really want to check whether the index was created in the ES cluster? Without that, how are we sure that this functionality works?

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java [hunk 2]
```diff
@@ -211,6 +211,22 @@ public void write(Collection<SinkRecord> records) {
...
       bulkProcessor.add(indexableRecord, flushTimeoutMs);
     }
+
+  }
+
+  /**
+   * Return the expected index name for a given topic. It makes sure to lowercase the topic name
+   * as elasticsearch require all index names to be lowercase.
+   * See https://github.com/confluentinc/kafka-connect-elasticsearch/issues/246 for connector details.
+   * See https://github.com/elastic/elasticsearch/issues/29420 for elasticsearch details.
```
Comment (rhauch) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: Minor: How about
```
   * Return the expected index name for a given topic, using the configured mapping or the topic name. Elasticsearch
   * <a href="https://github.com/elastic/elasticsearch/issues/29420">accepts only lowercase index names</a>.
 ```

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java [hunk 3]
```diff
@@ -211,6 +211,22 @@ public void write(Collection<SinkRecord> records) {
...
+   * @param topic The topic name being proceed.
+   * @return String A valid elasticsearch index name
+   */
+  private String convertTopicToIndexName(String topic) {
+    final String indexOverride = topicToIndexMap.get(topic);
+    String index = indexOverride != null ? indexOverride : topic.toLowerCase();
```
Comment (rhauch) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]:
```suggestion
    String index = indexOverride != null ? indexOverride : topic.toLowerCase(StandardCharsets.UTF_8);
```

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java [hunk 4]
```diff
@@ -211,6 +211,22 @@ public void write(Collection<SinkRecord> records) {
...
+    log.debug("Topic " + topic + " was translated as index name " + index);
```
Comment (rhauch) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: The following will avoid having to create the string when debug is not enabled:
```suggestion
    log.debug("Topic '{}' was translated as index '{}'", topic, index);
```

The challenge, of course, is that this log message is written out for every record. What do you think about keeping another cache (map) of the topic names to index name that were used so far, and then logging this message at INFO level when the topic had not yet been seen? It's a bit more complicated, but it would be much better to have that INFO level log message. (The cache really is more for bookkeeping than for performance.)

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java (hunk 1 above)
Comment (purbon) [reply src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java]: If you try to create an index with caps an exception will be throw, should look something like:

```java
org.apache.kafka.connect.errors.ConnectException: Could not create index 'AnotherTopicInCaps': {"root_cause":[{"type":"invalid_index_name_exception","reason":"Invalid index name [AnotherTopicInCaps], must be lowercase","index":"AnotherTopicInCaps"}],"type":"invalid_index_name_exception","reason":"Invalid index name [AnotherTopicInCaps], must be lowercase","index":"AnotherTopicInCaps"}
	at __randomizedtesting.SeedInfo.seed([A1BAC01989C87C:D2B7A48645E95AFE]:0)
	at io.confluent.connect.elasticsearch.ElasticsearchWriter.createIndicesForTopics(ElasticsearchWriter.java:268)
	at io.confluent.connect.elasticsearch.ElasticsearchSinkTask.open(ElasticsearchSinkTask.java:120)
	at io.confluent.connect.elasticsearch.ElasticsearchSinkTaskTest.testPutWithTopicInCaps(ElasticsearchSinkTaskTest.java:106)
```
I've learn this pattern over the year, probably contaminated with ruby and other languages, so yes to answer your question. I now the assertTrue does nothing, is mostly to show the message, the real check for this to work in on the fail part.

I am happy to write it more politely and learn, please let me know how it should look? should I just remove the assert?

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java (hunk 2 above)
Comment (purbon) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: fair for me.

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java (hunk 3 above)
Comment (purbon) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: lowercase accepts locale, but not Charset, right? we might like to add this kind of encoding validation before we reach here, isn't?

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java (hunk 4 above)
Comment (purbon) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: re the cache, something to think about it, certainly a good idea, but I think for this case and how we create indexes right now (only at bootstrap time), we are good the way we are. what do you think?

I am thinking for when we allow for an unseen index/topic to be created at the ES side.

Comment (purbon) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: thanks for the suggestion, happy to learn better ways of doing work.

**Comment by purbon** [comment]
Hi @rhauch, did you had any chance to take a look to the changes I recently introduced to this PR? Looking forward to know your thoughts.

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java (hunk 3 above)
Comment (rhauch) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: Oh, sorry. My bad. And this is actually not required, so it's okay to not change this line.

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java (hunk 1 above)
Comment (purbon) [reply src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java]: Very useful feedback @rhauch, pushing again my java testing skills 😄 forward, love that!

**Comment by purbon** [comment]
@rhauch updated the PR with your last feedaback, looking forward to green light from you 😄

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java [hunk 5]
```diff
@@ -189,9 +189,9 @@ public ElasticsearchWriter build() {
   }
 
   public void write(Collection<SinkRecord> records) {
+
```
Comment (rhauch) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: We generally try to avoid unnecessary changes like adding/removing lines so that the diff is smaller and more focused.
```suggestion
```

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java [hunk 6]
```diff
@@ -211,6 +211,19 @@ public void write(Collection<SinkRecord> records) {
 
       bulkProcessor.add(indexableRecord, flushTimeoutMs);
     }
+
```
Comment (rhauch) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: Let's not introduce this blank line unnecessarily.
```suggestion
```

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java [hunk 7]
```diff
@@ -211,6 +211,19 @@ public void write(Collection<SinkRecord> records) {
...
+  }
+
```
Comment (rhauch) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchWriter.java]: Let's not introduce this extra blank line unnecessarily. We prefer 1 blank line between methods.
```suggestion
```

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java [hunk 8]
```diff
@@ -72,4 +79,38 @@ public void testPutAndFlush() throws Exception {
...
+    try {
+      task.start(props, client);
+      task.open(new HashSet<>(Collections.singletonList(TOPIC_IN_CAPS_PARTITION)));
+      task.put(Collections.singleton(sinkRecord));
+    } catch (Exception ex) {
+      fail("A topic name not in lowercase can not be used as index name in Elasticsearch");
+    } finally {
+      task.stop();
+    }
+
```
Comment (rhauch) [inline src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java]: Again, let's remove unnecessary blank lines.
```suggestion
```

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java [hunk 9]
```diff
@@ -72,4 +79,38 @@ public void testPutAndFlush() throws Exception {
...
+  }
+
```
Comment (rhauch) [inline src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java]: One more: let's remove unnecessary blank lines.
```suggestion
```

**Comment by purbon** [comment]
@rhauch applyed your last request of changes at 72f996ec3090fb1078cd8913bd9acf90fd04b901
